File: apps/clojure/design-pattern/src/design_pattern/strategy.clj
(ns design-pattern.strategy)

;; Strategy パターン
;; Clojure では関数が第一級オブジェクトなので、戦略は単なる関数。

(defn format-report
  "レポートデータを受け取り、formatter 関数で整形する。"
  [formatter report]
  (formatter report))

;; --- HTML 戦略 ---

(defn html-formatter [{:keys [title items]}]
  (str "<html><head><title>" title "</title></head><body>\n"
       (apply str (map #(str "  <p>" % "</p>\n") items))
       "</body></html>\n"))

;; --- Plain Text 戦略 ---

(defn text-formatter [{:keys [title items]}]
  (str "=== " title " ===\n"
       (apply str (map #(str "  - " % "\n") items))
       "==========\n"))

;; --- Markdown 戦略 ---

(defn markdown-formatter [{:keys [title items]}]
  (str "# " title "\n\n"
       (apply str (map #(str "- " % "\n") items))))

File: apps/go/design-pattern/strategy/strategy.go
// Package strategy demonstrates the Strategy pattern in Go.
// Functions are first-class values in Go, so we use a function type
// as the strategy interface.
package strategy

import "strings"

// Formatter is a strategy function that formats a report.
type Formatter func(title string, text []string) string

// Report holds a title, text lines, and a formatting strategy.
type Report struct {
	Title     string
	Text      []string
	Formatter Formatter
}

// Output applies the formatter strategy to produce the report string.
func (r *Report) Output() string {
	if r.Formatter == nil {
		return ""
	}
	return r.Formatter(r.Title, r.Text)
}

// HtmlFormatter formats a report as HTML.
func HtmlFormatter(title string, text []string) string {
	var lines []string
	lines = append(lines, "<html>")
	lines = append(lines, " <head>")
	lines = append(lines, "  <title>"+title+"</title>")
	lines = append(lines, " </head>")
	lines = append(lines, "<body>")
	for _, t := range text {
		lines = append(lines, "  <p>"+t+"</p>")
	}
	lines = append(lines, "</body>")
	lines = append(lines, "</html>")
	return strings.Join(lines, "\n")
}

// PlainTextFormatter formats a report as plain text.
func PlainTextFormatter(title string, text []string) string {
	var lines []string
	lines = append(lines, "**** "+title+" ****")
	lines = append(lines, "")
	for _, t := range text {
		lines = append(lines, t)
	}
	return strings.Join(lines, "\n")
}
